### PR TITLE
feat(datasets) Simplify join structure

### PR DESCRIPTION
**Pull request by fpacifici**
While I was adding the FROM clause abstraction to the Query class I realized that the join tree structure could be simplified and that I could improve the naming of the classes. This change was orthogonal with respect to the Query changes so I extracted it into its own PR

test plan:
- the groups dataset works
- all the tests pass

### CODE REVIEW

File: snuba/datasets/schemas/join.py [hunk 1]
```diff
@@ -75,28 +76,7 @@ def get_schemas(self) -> Mapping[str, Schema]:
 
 
 @dataclass(frozen=True)
-class SubJoinSource(JoinedSource):
```
Comment (fpacifici) [inline snuba/datasets/schemas/join.py]: I think this had a reason to exist at the beginning, but after several rounds of refactoring I missed the fact it did not bring anything anymore.

File: snuba/datasets/schemas/join.py [hunk 2]
```diff
@@ -106,37 +86,47 @@ class JoinStructure:
...
 
     def __init__(self,
-        join_root: JoinStructure,
+        join_root: JoinNode,
     ) -> None:
         super().__init__(
             columns=self.__get_columns(join_root),
         )
-        self.__join_structure = join_root
+        self.__source = join_root
```
Comment (lynnagara) [inline snuba/datasets/schemas/join.py]: Is this used somewhere?

**Review by lynnagara — COMMENTED** [review]
Seems like a nice improvement

File: snuba/datasets/schemas/join.py (hunk 2 above)
Comment (fpacifici) [reply snuba/datasets/schemas/join.py]: Now yes, thanks for noticing

**Review by tkaemming — APPROVED** [review]

**Pull request merged**